File: 백준/Gold/2252. 줄 세우기/줄 세우기.py
from collections import deque
import sys

input = sys.stdin.readline
N, M = map(int, input().strip().split()) # 노드 수, 간선 수

tree = [[] for _ in range(N+1)]
in_degree = [0] * (N+1) # 노드별 입력차수
answer = []

for _ in range(M):
    a, b = map(int, input().strip().split()) # a -> b 방향
    tree[a].append(b)
    in_degree[b] += 1 # b 노드의 입력차수++

queue = deque([])
for i in range(1, N+1): # 1부터 N까지 입력차수 0인 노드 큐에 추가
    if not in_degree[i]:
        queue.append(i)

while queue:
    curr = queue.popleft() # 현재 노드
    answer.append(curr)
    for i in tree[curr]: # 인접노드와 연결 끊기(간선 제거)
        in_degree[i] -= 1 # 연결 노드의 입력차수--
        if not in_degree[i]: # 입력차수가 0이 되면 큐에 추가
            queue.append(i)

print(*answer)

File: 백준/Gold/2252. 줄 세우기/줄 세우기.java
import java.util.*;
import java.io.*;

class Main {
    public static void main(String[] args) throws IOException {
        
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken()); // 학생 수
        int M = Integer.parseInt(st.nextToken()); // 줄 세운 횟수

        List<Integer>[] tree = new ArrayList[N+1];
        int[] inDegree = new int[N+1]; // 노드별 입력차수
        List<Integer> answer = new ArrayList<>();
        
        for(int i = 1; i < N+1; i++) {
            tree[i] = new ArrayList<>();
        }
        
        for(int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            // 학생 A가 B앞에 옴
            int A = Integer.parseInt(st.nextToken());
            int B = Integer.parseInt(st.nextToken());

            tree[A].add(B);
            inDegree[B]++; // b노드의 입력차수++
        }
        br.close();

        Queue<Integer> queue = new LinkedList<>();
        for(int i = 1; i < N+1; i++) {
            if(inDegree[i] == 0) {
                queue.add(i); // 입력차수가 0인 노드만 추가
            }
        }

        while(!queue.isEmpty()) {
            int curr = queue.poll(); // 현재 노드
            answer.add(curr);

            // 현재 노드와 연결된 노드 탐색
            for(int adjNode : tree[curr]) {
                inDegree[adjNode]--;
                if(inDegree[adjNode] == 0) {
                    queue.add(adjNode);
                }
            }
        }

        for(int node : answer) {
            bw.write(node + " ");
        }
        bw.flush();
        bw.close();
    }
}